Userland/Libraries/LibVirtGPU/CommandBufferBuilder.cpp

```diff
@@ -0,0 +1,310 @@
+}
+
+void CommandBufferBuilder::append_create_vertex_elements(Protocol::ObjectHandle handle)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::CREATE_OBJECT, to_underlying(Protocol::ObjectType::VERTEX_ELEMENTS));
+    builder.appendu32(handle.value());
+    builder.appendu32(12); // src_offset_0
+    builder.appendu32(0);  // instance_divisor_0
+    builder.appendu32(0);  // vertex_buffer_index_0
+    builder.appendu32(30); // src_format_0 (PIPE_FORMAT_R32G32B32_FLOAT = 30)
+    builder.appendu32(0);  // src_offset_1
+    builder.appendu32(0);  // instance_divisor_1
+    builder.appendu32(0);  // vertex_buffer_index_1
+    builder.appendu32(30); // src_format_1 (PIPE_FORMAT_R32G32B32_FLOAT = 30)
+}
+
+void CommandBufferBuilder::append_bind_vertex_elements(Protocol::ObjectHandle handle)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::BIND_OBJECT, to_underlying(Protocol::ObjectType::VERTEX_ELEMENTS));
+    builder.appendu32(handle.value()); // VIRGL_OBJ_BIND_HANDLE
+}
+
+void CommandBufferBuilder::append_create_surface(Protocol::ResourceID drawtarget_resource, Protocol::ObjectHandle drawtarget_handle, Protocol::TextureFormat format)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::CREATE_OBJECT, to_underlying(Protocol::ObjectType::SURFACE));
+    builder.appendu32(drawtarget_handle.value());
+    builder.appendu32(drawtarget_resource.value());
+    builder.appendu32(to_underlying(format));
+    builder.appendu32(0); // First element / Texture Level
+    builder.appendu32(0); // Last element / Texture Element
+}
+
+void CommandBufferBuilder::append_set_framebuffer_state(Protocol::ObjectHandle drawtarget, Protocol::ObjectHandle depthbuffer)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::SET_FRAMEBUFFER_STATE, 0);
+    builder.appendu32(1);                   // nr_cbufs
+    builder.appendu32(depthbuffer.value()); // zsurf_handle
+    builder.appendu32(drawtarget.value());  // surf_handle
+}
+
+void CommandBufferBuilder::append_gl_viewport(Gfx::IntSize size)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::SET_VIEWPORT_STATE, 0);
+    builder.appendu32(0);
+    builder.appendf32(size.width() / 2);  // scale_x
+    builder.appendf32(size.height() / 2); // scale_y (flipped, due to VirGL being different from our coordinate space)
+    builder.appendf32(0.5f);              // scale_z
+    builder.appendf32(size.width() / 2);  // translate_x
+    builder.appendf32(size.height() / 2); // translate_y
+    builder.appendf32(0.5f);              // translate_z
+}
+
+void CommandBufferBuilder::append_set_framebuffer_state_no_attach(Gfx::IntSize size)
+{
+    CommandBuilder builder(m_buffer, Protocol::VirGLCommand::SET_FRAMEBUFFER_STATE_NO_ATTACH, 0);
+    builder.appendu32((size.height() << 16) | size.width()); // (height << 16) | width
+    builder.appendu32(0);                                    // (samples << 16) | layers
```

Comment:
First comment can be dropped, second comment can be converted into code (with `u16 samples = 0; u16 layers = 0;` somewhere)